refactor(agg): merge agg function implementations in batch and streaming

wangrunji0408 opened the pull request:
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR unifies the agg function implementations between batch and streaming.
After this PR, `agg_impl` and `minput_agg_impl` in streaming are removed and merged into `expr/agg`.

Now, the `Aggregator` trait is the only interface for all aggregate functions. The `update_multi` method now takes a `StreamChunk` instead of `DataChunk`, in which `+/U+` is for accumulation and `-/U-` is for retraction. There are also new methods added in `Aggregator` trait for state management, like `get_state`, `set_state` and `reset`.

The `#[aggregate]` macro is also adapted for streaming aggregation. Functions with a `retract: bool` argument at last are recognized as retractable functions. For example:
```rust
#[aggregate("count(*) -> int64", init_state = "0")]
fn count<T>(state: i64, _: T, retract: bool) -> i64 {
    if retract {
        state - 1
    } else {
        state + 1
    }
}
```

Other functions are append-only and will panic on retraction.

Previously we have some wrappers over `Aggregator` defined in `expr/agg`, including `distinct`, `orderby`, `projection` and `filter`. They are only suitable for batch execution, so this PR moves them into batch executors.

Changes in streaming agg executors include:
- combine `ops`, `columns` and `visibility` into a single `chunk` as much as possible.
- use `StreamChunk::project` (renamed from `reorder_columns`) as much as possible.
- move `arg_indices` out of agg state and project chunks before `apply_chunk` (except for minput state).

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [x] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)
  This PR doesn't change performance-critical code. But it's better to run a bench as it is a big refactoring.
<!-- To manually trigger a benchmark, please check out [Notion](https://www.notion.so/risingwave-labs/Manually-trigger-nexmark-performance-dashboard-test-b784f1eae1cf48889b2645d020b6b7d3). -->
- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

wangrunji0408 commented on src/expr/src/agg/array_agg.rs:
```diff
@@ -41,108 +41,42 @@ impl From<State> for ListValue {
-    }
-
-    #[tokio::test]
-    async fn test_array_agg_with_order() -> Result<()> {
```
note: this test is moved to `src/batch/src/executor/aggregation/orderby.rs`

wangrunji0408 commented on src/expr/src/agg/general.rs:
```diff
@@ -195,258 +219,401 @@ fn bool_or(state: bool, input: bool) -> bool {
 
 #[cfg(test)]
 mod tests {
```
note: unit tests in this file are merged from `src/stream/src/executor/aggregation/agg_impl/foldable.rs`.

stdrc commented on src/batch/src/executor/aggregation/orderby.rs:
```diff
@@ -83,42 +82,57 @@ impl Aggregator for ProjectionOrderBy {
 
     async fn update_multi(
         &mut self,
-        input: &DataChunk,
+        input: &StreamChunk,
         start_row_id: usize,
         end_row_id: usize,
     ) -> Result<()> {
         self.unordered_values.reserve(end_row_id - start_row_id);
         for row_id in start_row_id..end_row_id {
-            let (row, vis) = input.row_at(row_id);
+            let (op, row, vis) = input.row_at(row_id);
+            assert_eq!(op, Op::Insert, "only support append");
             if vis {
                 self.push_row(row)?;
             }
         }
         Ok(())
     }
```
What about adding another method for stream chunk? Like, `update_stream` or `update_retractable`, so that batch-specific aggregators can just `unimplemented` it. This also helps optimize `update_multi` performance if possible. Also, seeing `StreamChunk` in batch seems a little bit weird.

stdrc commented on src/common/src/array/data_chunk.rs:
```diff
@@ -746,6 +745,9 @@ impl DataChunkTestExt for DataChunk {
     fn from_pretty(s: &str) -> Self {
         use crate::types::ScalarImpl;
         fn parse_type(s: &str) -> DataType {
+            if let Some(s) = s.strip_suffix("[]") {
+                return DataType::List(Box::new(parse_type(s)));
+            }
```
What's this for?

wangrunji0408 replied:
to support list type in `DataChunk::from_pretty`.

stdrc commented on src/common/src/array/stream_chunk.rs:
```diff
@@ -70,9 +72,8 @@ pub type Ops<'a> = &'a [Op];
-    ops: Vec<Op>,
-
-    pub(super) data: DataChunk,
+    ops: Arc<[Op]>,
```
Can we use `Box<[Op]>` here?

wangrunji0408 replied:
Just wanted to make StreamChunk cheaply cloneable.

stdrc commented on src/stream/src/executor/backfill/utils.rs:
```diff
@@ -184,7 +184,7 @@ fn mark_chunk_inner(
     let (ops, columns, visibility) = chunk.into_inner();
     let mapped_columns = output_indices.iter().map(|&i| columns[i].clone()).collect();
-    StreamChunk::new(ops, mapped_columns, visibility)
+    StreamChunk::new(ops, mapped_columns, visibility.into_visibility())
```
Can we have `Into` here?

wangrunji0408 replied:
I plan to refactor it into the following 2 methods:
```rust
StreamChunk::new(
    ops: impl Into<Arc<[Op]>>,
    columns: Vec<ArrayRef>,
)
StreamChunk::with_visibility(
    ops: impl Into<Arc<[Op]>>,
    columns: Vec<ArrayRef>,
    visibility: impl Into<Vis>,
)
```
and eventually get rid of `Option<Bitmap>` and only use `Vis` for visibilities.

But this would cause a big change to the codebase. So maybe next PR for sure?

stdrc reviewed (commented):
Still remain some files not reviewed, too long 🥵.

BTW, would you mind running some simple benchmark to demonstrate there's no performance regression?

wangrunji0408 commented:
performance report: http://risingwave-perf-test-dashboard-metabase.us-west-2.elasticbeanstalk.com/question/1777-nexmark-blackhole-compare?tag_a=nightly-20230717&tag_b=141938962ca1f6de3c1fe65934654aa90added32-x86_64&metrics=avg-source-output-rows-per-second

st1page commented on src/stream/src/executor/aggregation/agg_impl/foldable.rs:
```diff
@@ -1,1024 +0,0 @@
-
-    // TODO: refactor with macro
-    #[bench]
-    fn bench_foldable_sum_agg_without_vis(b: &mut Bencher) {
```
would you help to migrate and run these benchmarks?

wangrunji0408 replied:
They have been migrated to `src/expr/src/agg/general.rs`. I'm running these benches and will show the results soon.

wangrunji0408 replied:
The result is shocking. 🤯

  | Before (ns) | After(ns) | Boost(Nx)
-- | -- | -- | --
count_agg_vis_rate_0_05 | 6,021 | 796 | 8
count_agg_vis_rate_0_25 | 32,725 | 962 | 34
count_agg_vis_rate_0_5 | 72,935 | 1,249 | 58
count_agg_vis_rate_0_75 | 110,661 | 1,491 | 74
count_agg_without_vis | 78,730 | 1,561 | 50
max_agg_vis_rate_0_05 | 12,234 | 791 | 15
max_agg_vis_rate_0_25 | 60,977 | 1,018 | 60
max_agg_vis_rate_0_5 | 122,474 | 1,328 | 92
max_agg_vis_rate_0_75 | 182,863 | 1,615 | 113
max_agg_without_vis | 146,554 | 1,724 | 85
min_agg_vis_rate_0_05 | 12,241 | 795 | 15
min_agg_vis_rate_0_25 | 60,960 | 1,019 | 60
min_agg_vis_rate_0_5 | 121,962 | 1,327 | 92
min_agg_vis_rate_0_75 | 182,723 | 1,615 | 113
min_agg_without_vis | 146,539 | 1,723 | 85
sum_agg_vis_rate_0_05 | 15,333 | 852 | 18
sum_agg_vis_rate_0_25 | 66,860 | 1,439 | 46
sum_agg_vis_rate_0_5 | 131,628 | 2,365 | 56
sum_agg_vis_rate_0_75 | 200,731 | 3,065 | 65
sum_agg_without_vis | 283,170 | 3,418 | 83

wangrunji0408 replied:
Oops, it's not right. The old benches run 100 times in each iteration. 😅

st1page replied:
So exciting. :exploding_head:

wangrunji0408 replied:
Fixed and updated result. The tide is turned. I should fix the regression now. 🥵

st1page replied:
I think it is because of the slow iteration on bitmap

st1page replied:
https://github.com/risingwavelabs/risingwave/pull/9310

wangrunji0408 replied:
Right. It should use `iter_ones` in `update_multi`.

wangrunji0408 replied:
Result updated after using `iter_ones`.

A notable overhead is the `start_row_id..end_row_id` range filter in `update_multi`. This is only used for batch agg (hashagg: update single row at a time, sortagg: update a range at a time). Other usages are all full ranges. I'm considering whether we should remove this range from the `update` method, and refactor batch usages by setting visibilities.

stdrc replied:
> I'm considering whether we should remove this range from the update method, and refactor batch usages by setting visibilities.

+1 if performance is ok. This also helps increase code consistency.

wangrunji0408 replied:
But the performance is a concern. Rewriting `update_single` to `update_multi` with one bit set in visibility will definitely hurt performance. Perhaps we should first make batch hashagg update a group at a time (like streaming does).

wangrunji0408 replied:
Anyway, do you think the performance shown above is acceptable to get this PR merged? We can leave further optimization to the future.

stdrc replied:
> Rewriting `update_single` to `update_multi` with one bit set in visibility will definitely hurt performance. Perhaps we should first make batch hashagg update a group at a time (like streaming does).

I think we can get rid of `update_single` finally, do batch updating in batch hash agg just like in streaming, but not necessarily in this PR.

st1page replied:
> But the performance is a concern. Rewriting update_single to update_multi with one bit set in visibility will definitely hurt performance. Perhaps we should first make batch hashagg update a group at a time (like streaming does).

+1 for this.
maybe we should do different interfaces for the predicate. one is for range, one is for visibity

stdrc commented on src/stream/src/executor/hash_agg.rs:
```diff
@@ -134,7 +133,7 @@ struct ExecutionVars<K: HashKey, S: StateStore> {
 
     /// Cache for [`AggGroup`]s. `HashKey` -> `AggGroup`.
-    agg_group_cache: AggGroupCache<K, S>,
+    agg_group_cache: AggGroupCache<K>,
```
I would suggest keep the `Strategy` generic parameter here. Though the really difference happens when calling `build_change`, the strategy is a characteristic of the executor.

BugenZhao commented on src/expr/macro/src/lib.rs:
```diff
@@ -431,6 +431,8 @@ struct UserFunctionAttr {
     /// The last argument type is `&mut dyn Write`.
     write: bool,
+    /// The last argument type is `retract: bool`.
+    retract: bool,
```
What about making this a const-generic? 👀

wangrunji0408 replied:
It doesn't help for better performance and makes macro more complex. 🥵

  | retract: bool | Const RETRACT: bool | Cmp
-- | -- | -- | --
count_agg_vis_rate_0_05 | 146 | 147 | 1%
count_agg_vis_rate_0_25 | 443 | 430 | -3%
count_agg_vis_rate_0_5 | 858 | 844 | -2%
count_agg_vis_rate_0_75 | 1,187 | 1,192 | 0%
count_agg_without_vis | 950 | 1,083 | 14%
max_agg_vis_rate_0_05 | 156 | 164 | 5%
max_agg_vis_rate_0_25 | 490 | 513 | 5%
max_agg_vis_rate_0_5 | 926 | 982 | 6%
max_agg_vis_rate_0_75 | 1,306 | 1,385 | 6%
max_agg_without_vis | 1,080 | 954 | -12%
min_agg_vis_rate_0_05 | 156 | 164 | 5%
min_agg_vis_rate_0_25 | 471 | 498 | 6%
min_agg_vis_rate_0_5 | 926 | 956 | 3%
min_agg_vis_rate_0_75 | 1,305 | 1,398 | 7%
min_agg_without_vis | 1,081 | 952 | -12%
sum_agg_vis_rate_0_05 | 244 | 242 | -1%
sum_agg_vis_rate_0_25 | 858 | 855 | -0%
sum_agg_vis_rate_0_5 | 1,671 | 1,641 | -2%
sum_agg_vis_rate_0_75 | 2,598 | 2,447 | -6%
sum_agg_without_vis | 2,559 | 2,573 | 1%

BugenZhao commented on src/expr/macro/src/gen.rs:
```diff
@@ -328,46 +325,71 @@ impl FunctionAttr {
+        let args = (0..self.args.len()).map(|i| format_ident!("v{i}"));
+        let args = quote! { #(#args,)* };
+        let retract = match self.user_fn.retract {
+            true => quote! { matches!(input.ops()[row_id], Op::Delete | Op::UpdateDelete) },
```
What about using `get_unchecked` here?

wangrunji0408 replied:
This does help. 👍

  | ops[idx] | ops.get_unchecked | Cmp
-- | -- | -- | --
count_agg_vis_rate_0_05 | 152 | 146 | -4%
count_agg_vis_rate_0_25 | 462 | 443 | -4%
count_agg_vis_rate_0_5 | 905 | 858 | -5%
count_agg_vis_rate_0_75 | 1,279 | 1,187 | -7%
count_agg_without_vis | 949 | 950 | 0%
max_agg_vis_rate_0_05 | 162 | 156 | -4%
max_agg_vis_rate_0_25 | 510 | 490 | -4%
max_agg_vis_rate_0_5 | 1,001 | 926 | -7%
max_agg_vis_rate_0_75 | 1,425 | 1,306 | -8%
max_agg_without_vis | 1,080 | 1,080 | 0%
min_agg_vis_rate_0_05 | 162 | 156 | -4%
min_agg_vis_rate_0_25 | 526 | 471 | -10%
min_agg_vis_rate_0_5 | 1,017 | 926 | -9%
min_agg_vis_rate_0_75 | 1,419 | 1,305 | -8%
min_agg_without_vis | 1,106 | 1,081 | -2%
sum_agg_vis_rate_0_05 | 253 | 244 | -4%
sum_agg_vis_rate_0_25 | 929 | 858 | -8%
sum_agg_vis_rate_0_5 | 1,801 | 1,671 | -7%
sum_agg_vis_rate_0_75 | 2,886 | 2,598 | -10%
sum_agg_without_vis | 3,093 | 2,559 | -17%

stdrc replied:
Cool!

stdrc commented on src/stream/src/executor/aggregation/agg_group.rs:
```diff
@@ -189,20 +185,20 @@ impl<S: StateStore, Strtg: Strategy> EstimateSize for AggGroup<S, Strtg> {
         agg_calls: &[AggCall],
-        storages: &[AggStateStorage<S>],
-        result_table: &StateTable<S>,
+        storages: &[AggStateStorage<impl StateStore>],
+        result_table: &StateTable<impl StateStore>,
```
I think we'd better keep `S` here? to ensure they have the same StateStore type, and also can make code clean and consistent with other executors.

stdrc reviewed (commented):
Generally LGTM! Except the performance concern🥵

wangrunji0408 commented on src/stream/src/executor/aggregation/minput.rs:
```diff
@@ -218,14 +211,25 @@ impl<S: StateStore> MaterializedInputState<S> {
-        Ok(self.cache.get_output())
+
+        if let Some(aggregator) = &mut self.aggregator {
+            let chunks = self.cache.output_batches().collect_vec();
+            for chunk in chunks {
+                aggregator.update(&chunk).await?;
+            }
```
Can't pass compile due to `!Send` issue. 🥵

stdrc replied:
What if `fn output_batches(&self, chunk_size: usize) -> Box<dyn Iterator<Item = StreamChunk> + Send + '_>`🥵

wangrunji0408 commented on src/stream/src/executor/aggregation/agg_state_cache.rs:
```diff
@@ -114,90 +62,132 @@ pub trait AggStateCacheFiller {
+        let mut values = self.state_cache.values();
+        Box::new(std::iter::from_fn(move || {
+            // build data chunk from rows
+            const CHUNK_SIZE: usize = 1024;
```
Done. But simply moved the const CHUNK_SIZE to the caller side.

wangrunji0408 commented:
Update performance report: Q17 throughput -15% 🥵

http://risingwave-perf-test-dashboard-metabase.us-west-2.elasticbeanstalk.com/question/1777-nexmark-blackhole-compare?tag_a=nightly-20230720&tag_b=411f7d95c680834d7a0e4973fdf915ed3134488b-x86_64&metrics=avg-source-output-rows-per-second

st1page commented:
FYI: https://www.notion.so/risingwave-labs/Flamegraph-Generation-Manual-c6b866ab13984980a964c13f066b998b#ff7dbcff80e14632bb08547d235eab31

wangrunji0408 commented:
I found the regression was in `AggGroup::create` and  `AggGroup::apply_chunk`. I made several optimizations to relieve them:
1. Changed columns type in `DataChunk` to `Arc<[ArrayRef]>` so that chunks can be cheaply cloneable. 81d65eb
2. Avoided `futures::try_join_all` on `AggGroup::create`. Because the error type is too large and it caused large `memmove` overhead.  e7e458a
3. Removed range filter from `update_multi`.

Now the performance on nexmark Q17 is finally outperformed than the main branch.

http://risingwave-perf-test-dashboard-metabase.us-west-2.elasticbeanstalk.com/question/1777-nexmark-blackhole-compare?tag_a=nightly-20230727&tag_b=858503851f3c81e6fd2c1748cba56c4fce6e3612-x86_64&metrics=avg-source-output-rows-per-second

The micro-benches also show a better result (vs main):

  | Before (ns) | After(ns) | Cmp
-- | -- | -- | --
count_agg_vis_rate_0_05 | 63 | 114 | 81%
count_agg_vis_rate_0_25 | 267 | 482 | 81%
count_agg_vis_rate_0_5 | 578 | 893 | 54%
count_agg_vis_rate_0_75 | 853 | 1,302 | 53%
count_agg_without_vis | 963 | 217 | -77%
max_agg_vis_rate_0_05 | 120 | 99 | -18%
max_agg_vis_rate_0_25 | 601 | 470 | -22%
max_agg_vis_rate_0_5 | 1,198 | 844 | -30%
max_agg_vis_rate_0_75 | 1,795 | 1,215 | -32%
max_agg_without_vis | 1,597 | 1,009 | -37%
min_agg_vis_rate_0_05 | 120 | 101 | -16%
min_agg_vis_rate_0_25 | 600 | 466 | -22%
min_agg_vis_rate_0_5 | 1,201 | 843 | -30%
min_agg_vis_rate_0_75 | 1,795 | 1,214 | -32%
min_agg_without_vis | 1,602 | 1,010 | -37%
sum_agg_vis_rate_0_05 | 186 | 168 | -10%
sum_agg_vis_rate_0_25 | 884 | 689 | -22%
sum_agg_vis_rate_0_5 | 1,902 | 1,365 | -28%
sum_agg_vis_rate_0_75 | 3,085 | 2,125 | -31%
sum_agg_without_vis | 3,396 | 672 | -80%

st1page reviewed (approved):
LGTM, PTAL later @stdrc

BugenZhao commented:
> 1. Changed columns type in `DataChunk` to `Arc<[ArrayRef]>` so that chunks can be cheaply cloneable. [81d65eb](https://github.com/risingwavelabs/risingwave/commit/81d65eb9540db582e199caf13cf8e621bec20a14)

Do you think it's worthwhile to have a `DataChunkRef` type? 👀

wangrunji0408 commented:
> > 1. Changed columns type in `DataChunk` to `Arc<[ArrayRef]>` so that chunks can be cheaply cloneable. [81d65eb](https://github.com/risingwavelabs/risingwave/commit/81d65eb9540db582e199caf13cf8e621bec20a14)
> 
> Do you think it's worthwhile to have a `DataChunkRef` type? 👀

I have considered several times. Not sure if it's worth adding more complexity to the type system to improve performance. 🤔

wangrunji0408 commented:
> Removed range filter from update_multi.

As imagined, this caused +30% regression in batch hash agg. 🤡
Looks like we have to keep `update_multi` for batch.

st1page commented on src/expr/src/agg/mod.rs:
```diff
@@ -25,48 +28,44 @@ mod def;
+    /// This is used in streaming aggregation.
+    async fn update(&mut self, input: &StreamChunk) -> Result<()>;
+
+    /// Update the aggregator with a range of rows.
+    ///
+    /// This is used in batch aggregation.
+    async fn update_range(&mut self, input: &StreamChunk, range: Range<usize>) -> Result<()>;
```
1. just to re-confirm, the current `update_range` interface will consider both `range` and `input.vis`, not assuming the stream chunk is compacted(no visibility bitmap)?
2. I think the comments here are not the essential distinction, BatchHashAgg can use the `update` interface with visibility bitmap, and in some cases(such as EOWC) streaming executor can use the `update_range` too.

stdrc replied:
1. Yes

wangrunji0408 replied:
1. Yes
2. Right. should be rephrased as "mainly used in"

stdrc commented on src/stream/src/executor/aggregation/agg_group.rs:
```diff
@@ -185,46 +188,49 @@ impl EstimateSize for AggGroup {
-                    input_schema,
-                )
-            }))
+        // XXX(wrj): don't use try_join_all here to avoid weird memmove overhead
+        let mut states = Vec::with_capacity(agg_calls.len());
+        for (idx, agg_call) in agg_calls.iter().enumerate() {
+            let state = AggState::create(
+                agg_call,
+                &storages[idx],
+                prev_outputs.as_ref().map(|outputs| &outputs[idx]),
+                pk_indices,
+                group_key.as_ref(),
+                extreme_cache_size,
+                input_schema,
+            )
```
This can potentially affect performance of initialization/recovery if later introduced UDAF need to do IO on init?

wangrunji0408 replied:
Yes. This can be revert after #11275.

stdrc reviewed (approved):
LGTM

wangrunji0408 commented on src/common/src/array/data_chunk.rs:
```diff
@@ -57,7 +58,7 @@ use crate::util::value_encoding::{
 #[must_use]
 pub struct DataChunk {
-    columns: Vec<ArrayRef>,
+    columns: Arc<[ArrayRef]>,
```
Absolutely. In column reordering (projection) cases, the vector will be rebuilt but arrays are the same.

wangrunji0408 commented on src/stream/src/executor/aggregation/agg_state.rs:
```diff
@@ -85,13 +94,13 @@ impl AggState {
             }
             AggStateStorage::MaterializedInput { mapping, .. } => {
-                Self::MaterializedInput(MaterializedInputState::new(
+                Self::MaterializedInput(Box::new(MaterializedInputState::new(
```
Because `MaterializedInputState` is large. This is to keep `AggState` small.

wangrunji0408 commented:
😱
```
thread '<unnamed>' panicked at 'attempt to subtract with overflow', src/common/src/buffer/bitmap.rs:84:17
```
